api/v1alpha1/secretagentconfiguration_types.go

```diff
@@ -74,96 +74,159 @@ func init() {
 }
 
 // SecretConfig is the configuration for a specific Kubernetes secret
 type SecretConfig struct {
 	// +kubebuilder:validation:Required
-	Name      string `json:"name" yaml:"name,omitempty"`
-	Namespace string `json:"-" yaml:"namespace,omitempty"`
+	Name      string `json:"name"`
+	Namespace string `json:"-"`
+
+	// This flag tells the controller to generate the secret only if the controller can't find it in k8s or the secret manager.
+	// This is useful if the user wants to enforce the use of the provided secret and avoid ever generating new ones.
+	// Defaults to true.
+	GenerateIfNecessary *bool `json:"generateIfNecessary,omitempty"`
+
 	// +kubebuilder:validation:Required
 	// +kubebuilder:validation:MinItems=1
-	Keys []*KeyConfig `json:"keys" yaml:"keys,omitempty" validate:"dive,unique=Name"`
+	Keys []*KeyConfig `json:"keys" validate:"dive,unique=Name"`
 }
 
 // KeyConfig is the configuration for a specific data key
 type KeyConfig struct {
 	// +kubebuilder:validation:Required
-	Name string `json:"name" yaml:"name,omitempty"`
+	Name string `json:"name"`
 	// +kubebuilder:validation:Required
-	Type           KeyConfigType  `json:"type" yaml:"type,omitempty"`
-	Value          string         `json:"value,omitempty" yaml:"value,omitempty"`
-	Length         int            `json:"length,omitempty" yaml:"length,omitempty"`
-	CAPath         []string       `json:"caPath,omitempty" yaml:"caPath,omitempty"`
-	PrivateKeyPath []string       `json:"privateKeyPath,omitempty" yaml:"privateKeyPath,omitempty"`
-	StorePassPath  []string       `json:"storePassPath,omitempty" yaml:"storePassPath,omitempty"`
-	KeyPassPath    []string       `json:"keyPassPath,omitempty" yaml:"keyPassPath,omitempty"`
-	AliasConfigs   []*AliasConfig `json:"keystoreAliases,omitempty" yaml:"keystoreAliases,omitempty,omitempty"`
-	Node           *Node          `json:"-" yaml:"node,omitempty"`
+	Type KeyConfigType `json:"type"`
+	Spec *KeySpec      `json:"spec,omitempty"`
+
+	// TODO Delete all items below this line
+	Value          string         `json:"value,omitempty"`
+	Length         int            `json:"length,omitempty"`
+	CAPath         []string       `json:"caPath,omitempty"`
+	PrivateKeyPath []string       `json:"privateKeyPath,omitempty"`
+	StorePassPath  []string       `json:"storePassPath,omitempty"`
+	KeyPassPath    []string       `json:"keyPassPath,omitempty"`
+	AliasConfigs   []*AliasConfig `json:"keystoreAliases,omitempty"`
+	Node           *Node          `json:"-"`
 }
 
+// KeySpec is the configuration for each key
+type KeySpec struct {
+	Value                 string        `json:"value,omitempty"`
+	Algorithm             AlgorithmType `json:"algorithm,omitempty"`
+	CommonName            string        `json:"commonName,omitempty"`
+	SignedWithPath        string        `json:"signedWithPath,omitempty"`
+	StoreType             StoreType     `json:"storeType,omitempty"`
+	StorePassPath         string        `json:"storePassPath,omitempty"`
```

Comment:
I dont think StorePassPath is needed per key - just per keystore
Also- I think we do not use a unique keypass per key - so this could be optionally omitted